### ISSUE TITLE
Type of `x or x` incorrectly inferred as `Never`

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#3005]
Pyright version: v1.1.219

Code:

```py
class Foo:
    pass

def f(foo: Foo) -> None:
    x = foo or foo
    reveal_type(x)
```

The type of `x` is revealed as `Never`. I expected the type to be `Foo`.

Interestingly, if you define the `__bool__` method, this works as expected:
```py
class Foo:
    def __bool__(self) -> bool:
        return True

def f(foo: Foo) -> None:
    x = foo or foo
    reveal_type(x)  # Foo
```

_but_ if the `__bool__` always returns `True`, the issue repeats
```py
from typing import Literal

class Foo:
    def __bool__(self) -> Literal[True]:
        return True

def f(foo: Foo) -> None:
    x = foo or foo
    reveal_type(x)  # Never
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.220, which I just published. It will also be included in the next release of pylance.
